### ISSUE TITLE
Advisory format: affected only if a cargo feature is enabled

### ISSUE THREAD

**Issue by kornelski** [rustsec/advisory-db#1266]
The advisory format doesn't have ability to express that a crate may be vulnerable only if an optional Cargo feature is enabled (e.g. when `cargo add halfbaddep` is fine, but `cargo add halfbaddep --features=with_reckless_abandon` exposes a vulnerability). The `functions` key is usable only if the feature adds new public API, but not if the feature only changes behavior. 

Such filter would be very helpful for `cargo-audit` and similar tools, because when a vulnerability is only in an unused optional feature, it can avoid reporting false positive vulnerability alerts.

**Comment by Shnatsel** [comment]
We could use `cargo metadata` to discover features. Reimplementing the Cargo.toml parser and feature resolver is infeasible, since the format keeps evolving and adding non-trivial complications.

**Comment by kornelski** [comment]
Indeed, features are not explicit in either `Cargo.lock` nor `cargo metadata`.

However, `cargo tree` can show dependencies with their features:

```bash
cargo tree -e features
cargo tree -f 'package {p} has features {f}'
```

Parsing this is a bit janky, since it's not meant to be a machine-readable format, but the `-f` arg could make it reasonably robust.

**Comment by kornelski** [comment]
Another alternative: how about a condition that the vulnerable crate has to have a certain dependency enabled? This info is present in `Cargo.lock` directly (`package.dependencies`), and could be a proxy for enabled features when the features enable an optional dependency.

**Comment by tarcieri** [comment]
`cargo-tree` obtains that information by linking directly to `cargo`, so at that point we should do the same, at least optionally (since it would add considerably to build times).

The idea of checking a particular package's dependencies is an interesting one.

**Comment by pinkforest** [comment]
Tooling feature so moving this under rustsec/rustsec and linking here to pinned issue:
https://github.com/rustsec/advisory-db/issues/1410
